Improved Elixir optlib parser

dreamtigers opened the pull request:
* Took the branch `elixir-optlib` and rebased it to a most recent master.
* Added some spaces in `optlib2c` for a better looking C code.
* Improved the Elixir parser `elixir.ctags` ([sent a pull request to the upstream](https://github.com/mmorearty/elixir-ctags/pull/8) but made the same improvements here, in fewer commits for conciseness)
* Added a more exhaustive test case for the new parser, as requested in #1806.

masatake commented on makefiles/optlib2c_input.mak:
```diff
@@ -3,7 +3,8 @@ OPTLIB2C_INPUT = \
 	optlib/ctags-optlib.ctags		\
-	optlib/elm.ctags			\
+	optlib/elixir.ctags		\
+	optlib/elm.ctags		\
```
You should not touch elm.ctags.

dreamtigers replied:
Yes, this was a conflict when rebasing 09861b6 to `master`.
The tabs in the line `optlib/elm.ctags` were corrected in 806f296, but that commit happened _after_ 09861b6, which triggered the conflict.

I should have seen and corrected the tabs, tho.

dreamtigers replied:
(trying out Github's suggestions feature, sorry if it's messy)
```suggestion
	optlib/elixir.ctags		\
	optlib/elm.ctags		\
```

masatake commented on misc/optlib2c:
```diff
@@ -826,22 +826,22 @@ EOF
-	def->kindTable = $opts->{'Clangdef'}KindTable;
-	def->kindCount = ARRAY_SIZE($opts->{'Clangdef'}KindTable);
+	def->kindTable     = $opts->{'Clangdef'}KindTable;
+	def->kindCount     = ARRAY_SIZE($opts->{'Clangdef'}KindTable);
```
These changes about whitespace have impacts on ALL generated code under optlib/ directory.
Before adding Elixir related change, you should make a pull request for this change. I will merge it first.
The pull request should include all changes under optlib/*.c, too.

dreamtigers replied:
Pull request done. There wasn't any other change under optlib/*.c, except for `optlib/elixir.c` but I think that change remains here (please correct me if I'm wrong).

How do I fix the fact that pull request #2026 and commit ffd1e98e contain the same changes? Sorry for the mess!

masatake commented on Units/simple-elixir.d/input.ex:
```diff
@@ -1,18 +1,190 @@
+#   defexception [:message]
+#
+#   @impl true
+#   def exception(value) do
```
An anonymous tag is needed here.  ctags main part should provide such a feature. 
This is my job.

masatake replied:
See #2025.

masatake commented on Units/simple-elixir.d/input.ex:
```diff
@@ -1,18 +1,190 @@
+#
+# f functions (def ...)
+#
+defmodule MyString do
```
Just a question. defmodule can be nested?
If not, you can build a scope. I will make more advises later.

masatake replied:
See a test case in #2025. It is an example of how scope flags work.

dreamtigers replied:
Yes, defmodule can be nested.

masatake commented:
You can run the test with "make units LANGUAGES=Elixir" locally.

masatake commented:
Could you consider a ".b" test case for the TODO items?
See .b test in "Gathering test cases for known bugs" in http://docs.ctags.io/en/latest/units.html.

Some valuable studies about changing the regex engine used in u-ctags are in #1861.

dreamtigers commented:
I renamed the `simple-elixir.d` directory to `parser-elixir.r`. Divided the test case in smaller cases, so it was worth it to rename the directory (as specified in `docs/units.rst`, in the section 'Categories'). And fixed one of the TODO's where the parser wasn't generating the right tags for the word logical operators (and,or,not, etc...).

masatake commented on docs/news.rst:
```diff
@@ -52,6 +52,7 @@ The following parsers have been added:
 * Diff
 * DTD
 * DTS
+* Elixir *optlib*
```
Excellent.

masatake commented:
Thank you. I will take over this. I will merge this after squashing some commits to keep change history simple. I'm in busy now. So, it will take few days.

masatake commented:
Merged after rearranging the commits posted this PR. See #2034.
Thank you.

dreamtigers commented:
Thanks a lot for the support and improving the parser! Much appreciated!
